src/Storages/VirtualColumnUtils.cpp

```diff
@@ -144,26 +147,75 @@ NameSet getVirtualNamesForFileLikeStorage()
     return getCommonVirtualsForFileLikeStorage().getNameSet();
 }
 
-std::unordered_map<std::string, std::string> parseHivePartitioningKeysAndValues(const String & path)
+std::pair<ColumnPtr, ColumnPtr> parseHivePartitioningKeysAndValues(const String & path)
 {
-    std::string pattern = "([^/]+)=([^/]*)/";
-    re2::StringPiece input_piece(path);
-
-    std::unordered_map<std::string, std::string> key_values;
-    std::string key;
-    std::string value;
-    std::unordered_map<std::string, std::string> used_keys;
-    while (RE2::FindAndConsume(&input_piece, pattern, &key, &value))
+    static auto extractor = KeyValuePairExtractorBuilder()
+    .withItemDelimiters({'/'})
+    .withKeyValueDelimiter('=')
+    .build();
+
+    auto keys = ColumnString::create();
+    auto values = ColumnString::create();
+
+    // cutting the filename to prevent malformed filenames that contain key-value-pairs from being extracted
+    // not sure if we actually need to do that, but just in case. Plus, the previous regex impl took care of it
+    const auto last_slash_pos = path.find_last_of('/');
+
+    if (last_slash_pos == std::string::npos)
+    {
+        // nothing to extract, there is no path, just a filename
+        return std::make_pair(std::move(keys), std::move(values));
+    }
+
+    std::string_view path_without_filename(path.data(), last_slash_pos);
+
+    extractor->extract(path_without_filename, keys, values);
+
+    keys->validate();
+    values->validate();
+
+    std::unordered_set<StringRef> check_for_duplicates_uset;
+
+    for (std::size_t i = 0; i < keys->size(); i++)
+    {
+        const auto key = keys->getDataAt(i);
+        const auto [it, inserted] = check_for_duplicates_uset.insert(key);
+
+        if (!inserted)
+        {
+            throw Exception(
+                ErrorCodes::INCORRECT_DATA,
+                "Path '{}' to file with enabled hive-style partitioning contains duplicated partition key {}, only unique keys are allowed",
+                path,
+                key.toString());
+        }
+    }
+
+    return std::make_pair(std::move(keys), std::move(values));
+}
+
+
+std::optional<std::pair<String, String>> getKeyAndValuePairFromHiveKeysAndValues(const std::optional<std::pair<ColumnPtr, ColumnPtr>> & hive_keys_and_values_opt, const String & key)
+{
+    if (!hive_keys_and_values_opt)
     {
-        auto it = used_keys.find(key);
-        if (it != used_keys.end() && it->second != value)
```

Comment:
This is something worth discussing: is it really necessary to do best-effort (i.e, only considering it a problem if the values are different) parsing here? It seems to me it's more of an invalid case rrather than something we want to support